[ValueTracking] Add missing check for two-value PN recurrance matching

ivanradanov opened the pull request:
When InstTy is a type like IntrinsicInst which can have a variable number of arguments, we can encounter a case where Operation will have fewer than two arguments and error at the getOperand() calls.

I think this was introduced by [c11ea449e59cf](https://github.com/llvm/llvm-project/pull/145964)

An alternative fix would be to have a version of `matchTwoInputRecurrence` where we pass `I` to it [here](https://github.com/llvm/llvm-project/blob/82d633e9ff51962eb21fb58d38b74b4e8829a199/llvm/lib/Analysis/ValueTracking.cpp#L9198) and check `I == Operation` in the if instead of `Operation->getNumOperands() == 2` .

ivanradanov committed:
[ValueTracking] Add missing check for two-value PN recurrance matching

When InstTy is a type like IntrinsicInst which can have a variable number of
arguments, we can encounter a case where Opetaion will have fewer than two
arguments and error at the getOperand() calls.

nikic reviewed (commented):
Can you please add a test?

ivanradanov committed:
Actually check >= operands

ivanradanov commented:
I added a test. For reference, before the patch, this is how it crashes:
```
AnalysisTests: src/llvm-project/llvm/include/llvm/IR/InstrTypes.h:2353: Value *llvm::CallBase::getOperand(unsigned int) const: Assertion `i_nocapture < OperandTraits<CallBase>::operands(this) && "getOperand() out of range!"' failed.
```
I also noticed that the `== 2` does not work for intrinsics as they also have the callee argument so I changed to `>= 2`

ivanradanov commented on llvm/unittests/Analysis/ValueTrackingTest.cpp:
```diff
@@ -1310,6 +1310,37 @@ TEST_F(ValueTrackingTest, MatchBinaryIntrinsicRecurrenceNegativeFSHR) {
+      %i6 = tail call noundef i32 @llvm.umin.i32(i32 %i4, i32 %arg)
+      ret void
+    }
+  )");
```
Yes it is currently possible, but the reproducable-in-lit example is a very artificial one produced by llvm-reduce and its ability to test this specific bug will probably be very flaky depending on future changes in instcombine.

for reference this is how it looks like and removing anything more stops testing the change
```
; Function Attrs: nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write)
declare void @llvm.assume(i1 noundef) #0

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare noundef range(i32 1, -2147483648) i32 @llvm.nvvm.read.ptx.sreg.nctaid.x() #1

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare i32 @llvm.umin.i32(i32, i32) #1

define i32 @foo(i32 %arg) {
bb:
  br i1 false, label %bb1, label %bb2

bb1:                                              ; preds = %bb
  br label %bb3

bb2:                                              ; preds = %bb
  %i = tail call noundef range(i32 1, -2147483648) i32 @llvm.nvvm.read.ptx.sreg.nctaid.x()
  br label %bb3

bb3:                                              ; preds = %bb2, %bb1
  %i4 = phi i32 [ 0, %bb1 ], [ %i, %bb2 ]
  %i5 = icmp samesign ult i32 0, %i4
  tail call void @llvm.assume(i1 %i5)
  %i6 = tail call noundef i32 @llvm.umin.i32(i32 %i4, i32 %arg)
  %i7 = icmp samesign ult i32 0, %i6
  br i1 %i7, label %bb8, label %bb9

bb8:                                              ; preds = %bb3
  ret i32 0

bb9:                                              ; preds = %bb3
  ret i32 0
}

attributes #0 = { nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) }
attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

I am not sure if it's valuable adding it.

nikic replied:
Wouldn't this reliably reproduce the issue, even in a lit test?
```
define i32 @foo(i1 %c, i32 %arg) {
entry:
  %i = call i32 @llvm.nvvm.read.ptx.sreg.nctaid.x()
  br i1 %c, label %bb.1, label %bb.2
bb.1:
  br label %bb.2
bb.2:
  %phi = phi i32 [ %i, %entry ], [ 0, %bb.1 ]
  %res = call i32 @llvm.umin.i32(i32 %phi, i32 %arg)
  ret i32 %res
}
```

ivanradanov replied:
You are right, it does. Thank you very much, I added this instead of the unit test.

antoniofrighetto commented on llvm/unittests/Analysis/ValueTrackingTest.cpp:
```diff
@@ -1310,6 +1310,37 @@ TEST_F(ValueTrackingTest, MatchBinaryIntrinsicRecurrenceNegativeFSHR) {
+    declare noundef i32 @llvm.nvvm.read.ptx.sreg.nctaid.x()
+    declare i32 @llvm.umin.i32(i32, i32)
+
+    define void @foo(i32 %arg) {
+    bb:
+      br i1 false, label %bb1, label %bb2
+
+    bb1:                                              ; preds = %bb
+      br label %bb3
+
+    bb2:                                              ; preds = %bb
+      %i = tail call noundef i32 @llvm.nvvm.read.ptx.sreg.nctaid.x()
+      br label %bb3
+
+    bb3:                                              ; preds = %bb2, %bb1
+      %i4 = phi i32 [ 0, %bb1 ], [ %i, %bb2 ]
+      %i6 = tail call noundef i32 @llvm.umin.i32(i32 %i4, i32 %arg)
+      ret void
+    }
```
```suggestion
    define void @foo(i32 %arg) {
    entry:
      %i = tail call noundef i32 @llvm.nvvm.read.ptx.sreg.nctaid.x()
      br i1 false, label %bb.1, label %bb.2
    bb.1:                                            ; preds = %entry
      br label %bb.2
    bb.2:                                            ; preds = %bb.2, %bb.1
      %i4 = phi i32 [ %i, %entry ], [ 0, %bb.1 ]
      %i6 = tail call noundef i32 @llvm.umin.i32(i32 %i4, i32 %arg)
      ret void
    }
```

ivanradanov committed:
Add lit test

nikic reviewed (approved):
LGTM, thanks

ivanradanov committed:
Add checks
